### ISSUE TITLE
FS segment accessing bug

### ISSUE THREAD

**Issue by expend20** [cea-sec/miasm#992]
Hey! I'm trying to pass trough the MS VS startup CRT code with jitter. The issue is the jitter stuck at this instruction:
```
MOV        EAX, DWORD PTR FS:[0x18]
```
it gives: `WARNING: address 0x18 is not mapped in virtual memory`. But I actuaaly mapped FS segment with the code:
```python
FS_0_ADDR = 0x1ff70000
sb.jitter.cpu.FS = 0x4
sb.jitter.cpu.set_segm_base(sb.jitter.cpu.FS, FS_0_ADDR)
sb.jitter.vm.add_memory_page(
    FS_0_ADDR, PAGE_READ + PAGE_WRITE, "\x00" * 0x100,
    "FS segment")
```
And previous instruction is working with FS segment just fine, for example:
```
...
00401975 PUSH       DWORD PTR FS:[0x0]
EAX 73DCEE82 EBX 00000000 ECX 8C23117D EDX 00000000 ESI 00000000 EDI 00000000 ESP 0013FFDC EBP 00000000 EIP 0040197C zf 0 nf 0 of 0 cf 0 
...
004019AE MOV        DWORD PTR FS:[0x0], EAX
EAX 0013FFDC EBX 00000000 ECX 8C23117D EDX 00000000 ESI 00000000 EDI 00000000 ESP 0013FFB4 EBP 0013FFEC EIP 004019B4 zf 0 nf 0 of 0 cf 0 
...
```

I also tried this snippet with no luck:
```python
sb.jitter.ir_arch.do_stk_segm = True
sb.jitter.ir_arch.do_ds_segm = True
sb.jitter.ir_arch.do_str_segm = True
sb.jitter.ir_arch.do_all_segm = True
```

Here is the full script, and the [binary itself](https://transfer.sh/vGNWh/test2.rar)

```python
#!/usr/bin/env python

from pdb import pm
from miasm2.analysis.sandbox import Sandbox_Win_x86_32
from miasm2.analysis.dse import DSEPathConstraint
from miasm2.jitter.csts import PAGE_READ, PAGE_WRITE

# Insert here user defined methods

def kernel32_QueryPerformanceCounter(jitter):
    ret_ad, _ = jitter.func_args_stdcall(["lpPerformanceCount"])
    jitter.func_ret_stdcall(ret_ad, 0)

def kernel32_IsProcessorFeaturePresent(jitter):
    ret_ad, _ = jitter.func_args_stdcall(["ProcessorFeature"])
    jitter.func_ret_stdcall(ret_ad, 0)

def InitializeCriticalSectionEx(jitter):
    #api-ms-win-core-synch-l1-2-0_InitializeCriticalSectionEx
    ret_ad, _ = jitter.func_args_stdcall(["lpCriticalSection",
        "dwSpinCount", "Flags"])
    jitter.func_ret_stdcall(ret_ad, 0)

# Parse arguments
parser = Sandbox_Win_x86_32.parser(description="PE sandboxer")
parser.add_argument("filename", help="PE Filename")
options = parser.parse_args()

print("[+] running file %s at %s" % (options.filename, options.address))

g = globals()
g['api-ms-win-core-synch-l1-2-0_InitializeCriticalSectionEx'] = InitializeCriticalSectionEx

sb = Sandbox_Win_x86_32(options.filename, options, g)

sb.jitter.ir_arch.do_stk_segm = True
sb.jitter.ir_arch.do_ds_segm = True
sb.jitter.ir_arch.do_str_segm = True
sb.jitter.ir_arch.do_all_segm = True

FS_0_ADDR = 0x1ff70000
sb.jitter.cpu.FS = 0x4
sb.jitter.cpu.set_segm_base(sb.jitter.cpu.FS, FS_0_ADDR)
sb.jitter.vm.add_memory_page(
    FS_0_ADDR, PAGE_READ + PAGE_WRITE, "\x00" * 0x100,
    "FS segment")

print(sb.jitter.vm)

sb.jitter.set_trace_log(trace_instr=True, trace_regs=True, trace_new_blocks=True)

sb.run()

```
PS: I'm running it with `./script.py test2.exe`

**Comment by commial** [comment]
Hey,

So, two things here. First, indeed, you have to enable segment handling (by default, Miasm will not do the translation). That's why you have an access in 0x0 + 0x18 at `MOV        EAX, DWORD PTR FS:[0x18]`. And that's also why the bug is not exactly the same after enabling segment handling. As a side note, you can avoid enabling the option manually using Sandbox's option (try a `-h`, like `--usesegm`).
Second, once the segment are handled, Windows structures have to be simulated. For now, in Miasm, just some of them are implemented. That's why you need to set and add page for FS base. Again, it can also be done with Sandbox option (`--use-windows-structs`).
So, the:
```Python
sb.jitter.ir_arch.do_stk_segm = True
sb.jitter.ir_arch.do_ds_segm = True
sb.jitter.ir_arch.do_str_segm = True
sb.jitter.ir_arch.do_all_segm = True

FS_0_ADDR = 0x1ff70000
sb.jitter.cpu.FS = 0x4
sb.jitter.cpu.set_segm_base(sb.jitter.cpu.FS, FS_0_ADDR)
sb.jitter.vm.add_memory_page(
    FS_0_ADDR, PAGE_READ + PAGE_WRITE, "\x00" * 0x100,
    "FS segment")
```
part can be removed and replaced with `python script.py --usesegm --use-windows-structs test2.exe`.

Now, running the code, it ends on:
```
00401421 MOV        EAX, DWORD PTR FS:[0x18]
EAX 7FF70000 EBX 00000000 ECX 00000001 EDX 00000000 ESI 00000000 EDI 00000000 ESP 0013FFB0 EBP 0013FFEC EIP 00401427 zf 0 nf 0 of 0 cf 0
00401427 MOV        ESI, 0x403338
EAX 7FF70000 EBX 00000000 ECX 00000001 EDX 00000000 ESI 00403338 EDI 00000000 ESP 0013FFB0 EBP 0013FFEC EIP 0040142C zf 0 nf 0 of 0 cf 0
0040142C MOV        EDX, DWORD PTR [EAX + 0x4]
WARNING: address 0x7FF70004 is not mapped in virtual memory:
```
So it seems it is fetching the TEB linear address (`FS:[0x18]`), then the Stack base (`FS:[0x4]`) according to the TEB structure.
If you take a look at `build_teb` in `miasm2/os_dep/win_api_x86_32_seh.py`, which is the code in charge of initializing the TEB structure, you'll see it is actually building a sparse structure. Only a few fields are mapped in memory (in Miasm, the memory granularity is the byte), and the stack base is not part of it. You can also observe this behavior by dumping the memory layout:
```Python
$ python2 -i script.py test2.exe --usesegm --use-windows-structs
...
>>> sb.jitter.vm
Addr               Size               Access Comment
0x10000            0x77               RW_    Process environment
0x130000           0x10000            RW_    Stack
0x200000           0x4C               RW_    Process parameters
0x341EAC           0x10               RW_    Loader struct
0x341EC4           0x4                RW_    Loader struct dummy dllbase
0x342F00           0x34               RW_    Module info 'test2.exe'
0x343400           0x15               RW_    Module name 'test2.exe'
0x343500           0x15               RW_    Module path 'test2.exe'
0x401000           0x1000             R__    'test2.exe': '.text\x00\x00\x00'
0x402000           0x1000             R__    'test2.exe': '.rdata\x00\x00'
0x403000           0x1000             RW_    'test2.exe': '.data\x00\x00\x00'
0x7FF70000         0x4                RW_    TEB.NtTib.ExceptionList
0x7FF70018         0x4                RW_    TEB.NtTib.Self
0x7FF70030         0x4                RW_    TEB.ProcessEnvironmentBlock
0x7FFDF008         0x8                RW_    PEB
```

Long story short, you need to map at `tib_address + 4` 4 bytes, containing the value you want for the Stack base address.
Hope it helps

**Comment by expend20** [comment]
Wow, thanks a lot for the information, I understood it.

I also noticed that segmentation sandbox options is cached somehow, so when I change it, it still shows me previous error until reboot. Why this is happening? And how to clean that cache?

**Comment by commial** [comment]
Indeed, there is a cache to faster the JiT. It is located in /tmp/miasm_cache, you can safely remove the directory, it will clean the cache

**Comment by expend20** [comment]
Cache was the original issue, I noticed the single step logs does not change after script change, spent quite some time to figure out what is happening and created this ticket about FS segment in the end :)

**Comment by commial** [comment]
No problem, that's what issues are made for
